### ISSUE TITLE
`MemberNameEqualsClassName` false positive for a factory function with implicit return type

### ISSUE THREAD

**Issue by lwasyl** [detekt/detekt#4070]
The following code:

```kotlin
internal class Foo private constructor() {

    companion object {

        fun foo() = Foo()
    }
}
```

will fail with `MemberNameEqualsClassName` rule (`MemberNameEqualsClassName - [foo] at ...`). This error looks like a false positive, because the documentation says

> Factory functions that create an instance of the class are exempt from this rule.

and the _offending_ function does in fact create an instance of the class

## Your Environment

* Version of detekt used: 1.18

**Comment by cortinico** [comment]
Are you running with Type Resolution enabled? This is a rule with a "mixed behavior" (see https://github.com/detekt/detekt/issues/2994) based on you running with Type Resolution or not.

Specifically the code section that checks the factory method exception requires type resolution.

More on type resolution here: https://detekt.github.io/detekt/type-resolution.html

**Comment by lwasyl** [comment]
Yes, that was it, with `detektMain` (`detektTest` in my case) the error is not there 🙂

Does running Detekt with type resolution using Gradle incur some performance penalty related to that type resolution? I didn't find any mention in the links provided

**Comment by BraisGabin** [comment]
Yes, it does. It incurs in a notable performace penalty.

**Comment by cortinico** [comment]
> But I think that we should not allow false positives and force the usage of Type Resolution

Are you suggesting that we promote the rule to Type Resolution only?

**Comment by BraisGabin** [comment]
If there is not an easy way to fix it, yes.

**Comment by mateuszkwiecinski** [comment]
This is still reproducible in `v1.19.0-RC2`

**Comment by BraisGabin** [comment]
Thanks for the heads up. I forgot about this one. I just create a PR fixing this. It could introduce new false-negatives. But those are better than false-positives. If you want a the rule to work properly you need to enable type-solving.
